File: 438. Find All Anagrams in a String.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> ans = new ArrayList<>();
        if (s==null || s.length()<p.length()) {
            return ans;
        }
        int[] count = new int[26];
        for (char c : p.toCharArray()) {
            count[c-'a']++;
        }
        int start=0, end=0, pl=p.length(), diff=pl, sl=s.length();
        // build the window
        for (end=0; end<pl; end++) {
            char c = s.charAt(end);
            count[c-'a']--;
            if (count[c-'a']>=0) {
                diff--;
            }
        }
        if (diff==0) {
            ans.add(0);
        }
        while (end < sl) {
            char c = s.charAt(start);
            // if this is greater or equal to 0, this means that we need this char in the anagram
            if (count[c-'a']>=0) {
                diff++;    
            }
            count[c-'a']++;
            start++;
            c = s.charAt(end);
            count[c-'a']--;
            // if this is greater or equal to 0, this means that we need this char in the anagram
            if (count[c-'a']>=0) {
                diff--;
            }
            if (diff==0) {
                ans.add(start);
            }
            end++;
        }
        return ans;
    }
}

File: 438. Find All Anagrams in a String.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        if len(s) < len(p):
            return []
        pcount = [0] * 26
        for c in p:
            pcount[ord(c)-ord('a')] += 1
        diff = len(p)
        for i in range(len(p)):
            pcount[ord(s[i]) - ord('a')] -= 1
            if pcount[ord(s[i]) - ord('a')] >= 0:
                diff -= 1
        ans = []
        plen = len(p)
        if diff == 0:
            ans.append(0)
        for i in range(plen, len(s)):
            c = s[i-plen]
            pcount[ord(c) - ord('a')] += 1
            if pcount[ord(c) - ord('a')] > 0:
                diff += 1
            c = s[i]
            pcount[ord(c) - ord('a')] -= 1
            if pcount[ord(c) - ord('a')] >= 0:
                diff -= 1
            if diff == 0:
                ans.append(i-plen+1)
        return ans